front: add waypoint menu in manchette

SharglutDev opened the pull request:
close #8629 

Mockup : 
https://www.sketch.com/s/a26d84be-5fb2-41e4-9e84-96d90f46a997/a/OeJV9yG

SharglutDev committed:
front: add waypoint menu in manchette

- When clicking on a waypoint in the manchette, display a menu with some
actions
- Only action for now is to hide the waypoint in the manchette

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

emersion commented on front/src/common/OSRDMenu.tsx:
```diff
@@ -9,10 +9,11 @@ export type OSRDMenuItem = {
 type OSRDMenuProps = {
   menuRef: React.RefObject<HTMLDivElement>;
   items: OSRDMenuItem[];
+  style?: React.CSSProperties;
```
If this is exposed to allow callers to position the menu, should we instead add a `position` property? That would deter callers from (ab)using `style` for things which should stay in CSS files.

SharglutDev replied:
We also need to set the width, it's not always the same depending of who's calling the menu. I made the call to pass this props instead of pass `width` and `position`. You would prefer that ?

Synar replied:
I think it would be cleaner to specify width and position, this way callers know what they can and should provide

SharglutDev replied:
Forgot to update here but on the second fixup, I put the width in the css directly.

For the position part, I'll probably have to put the menu inside osrd-ui directly so we might not need this props anymore. I'll try that tomorrow.

emersion commented on front/src/modules/simulationResult/components/ManchetteWithSpaceTimeChart/ManchetteWithSpaceTimeChart.tsx:
```diff
@@ -7,10 +7,12 @@ import { ConflictLayer, PathLayer, SpaceTimeChart } from '@osrd-project/ui-space
 import type { Conflict } from '@osrd-project/ui-spacetimechart';
 
 import type { OperationalPoint, TrainSpaceTimeData } from 'applications/operationalStudies/types';
+import WaypointMenu from 'common/OSRDMenu';
```
Should this be named `OSRDMenu`? It's a bit confusing to rename on import when there are no naming conflicts.

SharglutDev replied:
Done

emersion commented on front/src/modules/simulationResult/components/ManchetteWithSpaceTimeChart/ManchetteWithSpaceTimeChart.tsx:
```diff
@@ -64,7 +85,19 @@ const ManchetteWithSpaceTimeChartWrapper = ({
+              menuRef={waypointMenuData.menuRef}
+              items={waypointMenuData.menuItems}
+              style={{
+                width: '305px',
```
Could we move this to a CSS file?

SharglutDev replied:
This would require a div wrapper as the width it's not always the same when calling the menu. I'm not a huge fan of that. Or you prefer passing a `className` prop instead of `style` ? How would you do ?

emersion replied:
Wouldn't `.manchette .osrd-menu` work? If not, I'd suggest adding a `className` prop.

SharglutDev replied:
Oh yes, you're right, done

emersion commented on front/src/modules/simulationResult/components/ManchetteWithSpaceTimeChart/ManchetteWithSpaceTimeChart.tsx:
```diff
@@ -41,8 +45,25 @@ const ManchetteWithSpaceTimeChartWrapper = ({
+      ...manchetteProps,
+      operationalPoints: manchetteProps.operationalPoints.map((op) => ({
+        ...op,
+        onClick: (id: string, ref: HTMLDivElement | null) => {
+          waypointMenuData.handleWaypointClick(id, ref);
+        },
```
Nit: can be simplified to:

```ts
onClick: waypointMenuData.handleWaypointClick,
```

SharglutDev replied:
Oh good point, done

emersion commented on front/src/modules/simulationResult/components/SpaceTimeChart/useWaypointMenu.tsx:
```diff
@@ -0,0 +1,71 @@
+        setMenuPosition(undefined);
+        setActiveOperationalPointId(undefined);
+        if (filteredWaypoints && setFilteredWaypoints) {
+          setFilteredWaypoints(
```
This is racy, because another update to `filteredWaypoints` could be pending. Instead, `setFilteredWaypoints` can take a callback which mutates the current value:

```ts
setFilteredWaypoints((filteredWaypoints) => filteredWaypoints.filter(…));
```

As a bonus, this removes the need to pass `filteredWaypoints` as a prop.

SharglutDev replied:
Yep I know but as it requires to pass the `setFilteredWaypoints` not as a standard function but as a `SetStateAction`, it needed to be changed everywhere we pass this props (but only two spots actually) and I was lazy :upside_down_face: But you're right I should do that, done :)

emersion commented on front/src/modules/simulationResult/components/SpaceTimeChart/useWaypointMenu.tsx:
```diff
@@ -0,0 +1,71 @@
+import useOutsideClick from 'utils/hooks/useOutsideClick';
+
+const SPACETIME_CHART_HEADER_HEIGHT = 40;
+const WAYPOINT_MENU_OFFSET = 2;
```
This could be a CSS margin instead.

SharglutDev replied:
Done

emersion commented on front/src/modules/simulationResult/components/SpaceTimeChart/useWaypointMenu.tsx:
```diff
@@ -0,0 +1,71 @@
+      title: t('waypointMenu.hide'),
+      icon: <EyeClosed />,
+      onClick: () => {
+        setMenuPosition(undefined);
+        setActiveOperationalPointId(undefined);
```
We could call `closeMenu` here instead.

SharglutDev replied:
Done

emersion reviewed (commented):
Overall I'm not a huge fan of having the menu outside of ui-manchette. IMHO the positioning and styling of the menu belongs to osrd-ui.

I'm especially concerned about the positioning logic which is quite fragile: it hardcodes a lot of assumptions about the manchette, and desynchronizes when elements move around. Example bug: zoom in the manchette, open the menu, and then scroll. This is just one example, other interactions may break the menu alike.

![out](https://github.com/user-attachments/assets/a76eda0e-e201-4161-b968-334ee6ba48c9)

emersion commented:
Actually, maybe this coule be used instead?

https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_anchor_positioning/Using

SharglutDev commented:
> Overall I'm not a huge fan of having the menu outside of ui-manchette. IMHO the positioning and styling of the menu belongs to osrd-ui.
> 
> I'm especially concerned about the positioning logic which is quite fragile: it hardcodes a lot of assumptions about the manchette, and desynchronizes when elements move around. Example bug: zoom in the manchette, open the menu, and then scroll. This is just one example, other interactions may break the menu alike.
> 

You're totally right, I will check for solutions

Synar reviewed (commented):
LGTM beyond the positioning issue pointed out by emersion. But it is my understanding that https://github.com/OpenRailAssociation/osrd-ui/pull/668/files should solve this issue?

SharglutDev commented:
> LGTM beyond the positioning issue pointed out by emersion. But it is my understanding that https://github.com/OpenRailAssociation/osrd-ui/pull/668/files should solve this issue?

Yes exactly it's on wip :)

SharglutDev commented:
Nothing new, just rebased with dev to be not too far behind with conflicts :)

SharglutDev committed:
front: fix useModalFocusTrap

- Change the listener to be on the document to be able to catch all events even if the focus was elsewhere
- Add a parameter to set the focus on the first element of the modal/div if provided
- Fix the focusTrap on the manchette menu

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

SharglutDev committed:
front: add disabled style to osrd menu items

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

clarani commented on front/src/modules/simulationResult/components/ManchetteWithSpaceTimeChart/ManchetteWithSpaceTimeChart.tsx:
```diff
@@ -178,8 +182,29 @@ const ManchetteWithSpaceTimeChartWrapper = ({
+  const waypointMenuData = useWaypointMenu(
+    waypointsPanelData?.filteredWaypoints,
+    waypointsPanelData?.setFilteredWaypoints
+  );
```
Nit: you can simply give the waypointsPanelData directly
```suggestion
  const waypointMenuData = useWaypointMenu(waypointsPanelData);
```

SharglutDev replied:
Done

theocrsb commented on front/src/styles/scss/common/components/_manchetteWithSpaceTimeChart.scss:
```diff
@@ -37,6 +38,15 @@
+
+    .osrd-menu {
+      width: 305px;
+      margin-top: -2px;
```
I think it's better not to use a negative margin, a transform would be better.

SharglutDev replied:
Done

theocrsb reviewed (approved):
LGTM and tested ! I just posted a little comment

clarani reviewed (approved):
LGTM :white_check_mark:
